File: android-catalogue/catalogue/src/main/java/xyz/ksharma/prisma/catalogue/playground/StatesPager.kt
package xyz.ksharma.prisma.catalogue.playground

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ColumnScope
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.pager.HorizontalPager
import androidx.compose.foundation.pager.rememberPagerState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import xyz.ksharma.prisma.coreui.themed
import xyz.ksharma.prisma.tokens.PrismaRadius
import xyz.ksharma.prisma.tokens.PrismaSemanticColors
import xyz.ksharma.prisma.tokens.PrismaSpacing
import xyz.ksharma.prisma.tokens.PrismaTypography

public data class PlaygroundState(
    public val label: String,
    public val content: @Composable () -> Unit,
)

/**
 * Horizontal pager of canonical component states. Page height is fixed —
 * crucial UX detail per feedback that variable-height pages caused jarring
 * layout shifts as the user swiped through. Each page renders inside a
 * card with a state label across the top.
 */
@Composable
public fun StatesPager(
    states: List<PlaygroundState>,
    modifier: Modifier = Modifier,
    pageHeight: Dp = 220.dp,
) {
    val pagerState = rememberPagerState(pageCount = { states.size })
    Column(
        modifier = modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(PrismaSpacing.Sp3),
    ) {
        // No contentPadding so each page is exactly the container's width —
        // matches the preview surface above so the two cards line up
        // edge-for-edge instead of looking like irregular shapes.
        HorizontalPager(
            state = pagerState,
            modifier = Modifier
                .fillMaxWidth()
                .height(pageHeight),
        ) { index ->
            val state = states[index]
            Column(
                modifier = Modifier
                    .fillMaxWidth()
                    .height(pageHeight)
                    .clip(RoundedCornerShape(PrismaRadius.Md))
                    .background(PrismaSemanticColors.SurfaceRaised.themed())
                    .border(
                        1.dp,
                        PrismaSemanticColors.BorderSubtle.themed(),
                        RoundedCornerShape(PrismaRadius.Md),
                    )
                    .padding(PrismaSpacing.Sp5),
                verticalArrangement = Arrangement.spacedBy(PrismaSpacing.Sp4),
            ) {
                Text(
                    text = state.label.uppercase(),
                    style = PrismaTypography.LabelSm,
                    color = PrismaSemanticColors.TextTertiary.themed(),
                )
                // Centered horizontally + vertically so a small button or
                // chip doesn't sit in the top-left corner of a 220dp cell —
                // the component is the focal point, not the label.
                Box(
                    modifier = Modifier.fillMaxWidth().weight(1f),
                    contentAlignment = Alignment.Center,
                ) {
                    state.content()
                }
            }
        }
        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.Center,
        ) {
            repeat(states.size) { idx ->
                val active = idx == pagerState.currentPage
                Box(
                    modifier = Modifier
                        .padding(horizontal = 3.dp)
                        .size(width = if (active) 18.dp else 6.dp, height = 6.dp)
                        .clip(CircleShape)
                        .background(
                            if (active) PrismaSemanticColors.AccentDefault.themed()
                            else PrismaSemanticColors.BorderDefault.themed(),
                        ),
                )
            }
        }
    }
}

File: ios-catalogue/App/Playground/StatesPager.swift
import SwiftUI
import CoreUI

public struct PlaygroundState<Content: View> {
    let label: String
    let content: () -> Content

    public init(_ label: String, @ViewBuilder content: @escaping () -> Content) {
        self.label = label
        self.content = content
    }
}

/// Horizontal pager of canonical component states. Page height is fixed —
/// crucial UX detail per feedback that variable-height pages caused
/// jarring layout shifts as the user swiped between states.
struct StatesPager: View {
    @Environment(\.colorScheme) private var scheme
    let states: [AnyPlaygroundState]
    var pageHeight: CGFloat = 220
    @State private var index: Int = 0

    var body: some View {
        VStack(spacing: PrismaSpacing.sp3) {
            TabView(selection: $index) {
                ForEach(Array(states.enumerated()), id: \.offset) { idx, state in
                    VStack(alignment: .leading, spacing: PrismaSpacing.sp4) {
                        Text(state.label.uppercased())
                            .font(PrismaTypography.labelSm.font)
                            .foregroundStyle(PrismaSemanticColors.textTertiary.themed(scheme))
                        state.content
                            .frame(maxWidth: .infinity, alignment: .leading)
                        Spacer(minLength: 0)
                    }
                    .padding(PrismaSpacing.sp5)
                    .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .topLeading)
                    .background(PrismaSemanticColors.surfaceRaised.themed(scheme))
                    .clipShape(RoundedRectangle(cornerRadius: PrismaRadius.md))
                    .overlay(
                        RoundedRectangle(cornerRadius: PrismaRadius.md)
                            .strokeBorder(PrismaSemanticColors.borderSubtle.themed(scheme), lineWidth: 1)
                    )
                    .padding(.horizontal, PrismaSpacing.sp4)
                    .tag(idx)
                }
            }
            .tabViewStyle(.page(indexDisplayMode: .never))
            .frame(height: pageHeight)

            HStack(spacing: 6) {
                ForEach(0..<states.count, id: \.self) { i in
                    let active = i == index
                    Capsule()
                        .fill(
                            active
                                ? PrismaSemanticColors.accentDefault.themed(scheme)
                                : PrismaSemanticColors.borderDefault.themed(scheme)
                        )
                        .frame(width: active ? 18 : 6, height: 6)
                        .animation(.easeInOut(duration: 0.2), value: index)
                }
            }
        }
    }
}

/// Type-erased playground state so a single-typed `[AnyPlaygroundState]`
/// array can mix differently-typed cell content.
public struct AnyPlaygroundState {
    let label: String
    let content: AnyView

    public init<Content: View>(_ label: String, @ViewBuilder content: () -> Content) {
        self.label = label
        self.content = AnyView(content())
    }
}
